src/timeout.ts

```diff
@@ -0,0 +1,129 @@
+import { clearTimeout, setTimeout } from 'timers';
+
+import { MongoError, MongoInvalidArgumentError } from './error';
+import { noop } from './utils';
+
+/** @internal */
+export class CSOTError extends MongoError {
+  override get name(): 'CSOTError' {
+    return 'CSOTError';
+  }
+
+  constructor(message: string, options?: { cause?: Error }) {
+    super(message, options);
+  }
+
+  static is(error: unknown): error is CSOTError {
+    return (
+      error != null && typeof error === 'object' && 'name' in error && error.name === 'CSOTError'
+    );
+  }
+
+  static from(error: CSOTError) {
+    return new CSOTError(error.message, { cause: error });
+  }
+}
+
+type Executor = ConstructorParameters<typeof Promise<never>>[0];
+type Reject = Parameters<ConstructorParameters<typeof Promise<never>>[0]>[1];
+/** @internal
```

Comment:
`@internal` needs to be on its own line, and this isn't CSOT specific anymore, we can update the docs.